Backend Lifecycle (Pass 2)

WhiredPlanck opened the pull request:
## Pull request

#### Issue tracker
Fixes will automatically close the related issues
<!-- Each issue should be on it's own line -->
Fixes # N/A

#### Feature
Describe features of this pull request

In this PR, things are done following:
1. Run the rime backend on a standalone single thread, making operations calling it can run async.
2. Migrate some API to run with the custom dispatcher context. 
3. By the way, the UX of select- / enabl-ing schema(ta) is significantly improved.

#### Code of conduct
- [x] [CONTRIBUTING](CONTRIBUTING.md)

#### Code style
- [x] `make sytle-lint`
- [x] [Conventional Commits](https://www.conventionalcommits.org/)

#### Build pass
- [x] `make debug`

#### Manually test
- [x] Done

#### Code Review
1. No wildcards import
4. Manual build and test pass
5. GitHub Action CI pass
6. At least one contributor review and approve
7. Merged clean without conflicts
8. PR will be merged by rebase upstream base

#### Daily build
Login and download artifact at https://github.com/osfans/trime/actions

#### Additional Info

WhiredPlanck committed:
refactor: replace handler with main looper with custom RimeLifecycleScope

WhiredPlanck committed:
refactor: add schemata stuffs to RimeApi

WhiredPlanck committed:
feat: implements RimeDispatcher for running the rime backend solo on a single thread

WhiredPlanck committed:
refactor: enhance the UX of selecting/enabling schema(ta)

- Apply new api implementations of schema
- More clear wording of the hints
- Refactors to the building of the picker dialogs

WhiredPlanck committed:
refactor: remove the judgment of whether the notification is handling

The native notification handler of rime is seems to be thread-safe now.

goofyz commented:
如刪除 `build` 的話，會
```
kotlin.UninitializedPropertyAccessException: lateinit property currentSchema has not been initialized
	at com.osfans.trime.data.schema.SchemaManager.getActiveSchema(SchemaManager.kt:26)
	at com.osfans.trime.ime.symbol.TabManager.addTabHasKeys(TabManager.kt:111)
	at com.osfans.trime.ime.symbol.TabManager.refresh(TabManager.kt:52)
	at com.osfans.trime.data.theme.ThemeManager.setNormalTheme(ThemeManager.kt:70)
	at com.osfans.trime.data.theme.ThemeManager.init(ThemeManager.kt:58)
	at com.osfans.trime.data.theme.ColorManager.init(ColorManager.kt:98)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invokeSuspend(TrimeInputMethodService.kt:238)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:8)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:4)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invokeSuspend(RimeDaemon.kt:59)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:8)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:4)
	at com.osfans.trime.core.RimeLifecycleKt.whenAtState(RimeLifecycle.kt:82)
	at com.osfans.trime.daemon.RimeDaemon$establish$1.runOnReady(RimeDaemon.kt:135)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1.invokeSuspend(TrimeInputMethodService.kt:236)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
	at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
	at com.osfans.trime.ime.core.TrimeInputMethodService.onCreate(TrimeInputMethodService.kt:235)
```	
	
或者

```
java.util.NoSuchElementException: No such color scheme (my_color_theme) found!
	at com.osfans.trime.data.theme.ColorManager.refresh(ColorManager.kt:119)
	at com.osfans.trime.data.theme.ThemeManager.setNormalTheme(ThemeManager.kt:69)
	at com.osfans.trime.data.theme.ThemeManager.init(ThemeManager.kt:58)
	at com.osfans.trime.data.theme.ColorManager.init(ColorManager.kt:98)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invokeSuspend(TrimeInputMethodService.kt:238)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:8)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:4)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invokeSuspend(RimeDaemon.kt:59)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:8)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:4)
	at com.osfans.trime.core.RimeLifecycleKt.whenAtState(RimeLifecycle.kt:82)
	at com.osfans.trime.daemon.RimeDaemon$establish$1.runOnReady(RimeDaemon.kt:135)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1.invokeSuspend(TrimeInputMethodService.kt:236)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
	at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
	at com.osfans.trime.ime.core.TrimeInputMethodService.onCreate(TrimeInputMethodService.kt:235)
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4485)
	at android.app.ActivityThread.-$$Nest$mhandleCreateService(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2160)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
```

WhiredPlanck commented:
> 如刪除 `build` 的話，會
> 
> ```
> kotlin.UninitializedPropertyAccessException: lateinit property currentSchema has not been initialized
> 	at com.osfans.trime.data.schema.SchemaManager.getActiveSchema(SchemaManager.kt:26)
> 	at com.osfans.trime.ime.symbol.TabManager.addTabHasKeys(TabManager.kt:111)
> 	at com.osfans.trime.ime.symbol.TabManager.refresh(TabManager.kt:52)
> 	at com.osfans.trime.data.theme.ThemeManager.setNormalTheme(ThemeManager.kt:70)
> 	at com.osfans.trime.data.theme.ThemeManager.init(ThemeManager.kt:58)
> 	at com.osfans.trime.data.theme.ColorManager.init(ColorManager.kt:98)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invokeSuspend(TrimeInputMethodService.kt:238)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:8)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:4)
> 	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invokeSuspend(RimeDaemon.kt:59)
> 	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:8)
> 	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:4)
> 	at com.osfans.trime.core.RimeLifecycleKt.whenAtState(RimeLifecycle.kt:82)
> 	at com.osfans.trime.daemon.RimeDaemon$establish$1.runOnReady(RimeDaemon.kt:135)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1.invokeSuspend(TrimeInputMethodService.kt:236)
> 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
> 	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
> 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
> 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
> 	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
> 	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
> 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
> 	at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
> 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
> 	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService.onCreate(TrimeInputMethodService.kt:235)
> ```
> 
> 或者
> 
> ```
> java.util.NoSuchElementException: No such color scheme (my_color_theme) found!
> 	at com.osfans.trime.data.theme.ColorManager.refresh(ColorManager.kt:119)
> 	at com.osfans.trime.data.theme.ThemeManager.setNormalTheme(ThemeManager.kt:69)
> 	at com.osfans.trime.data.theme.ThemeManager.init(ThemeManager.kt:58)
> 	at com.osfans.trime.data.theme.ColorManager.init(ColorManager.kt:98)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invokeSuspend(TrimeInputMethodService.kt:238)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:8)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:4)
> 	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invokeSuspend(RimeDaemon.kt:59)
> 	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:8)
> 	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:4)
> 	at com.osfans.trime.core.RimeLifecycleKt.whenAtState(RimeLifecycle.kt:82)
> 	at com.osfans.trime.daemon.RimeDaemon$establish$1.runOnReady(RimeDaemon.kt:135)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1.invokeSuspend(TrimeInputMethodService.kt:236)
> 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
> 	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
> 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
> 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
> 	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
> 	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
> 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
> 	at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
> 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
> 	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
> 	at com.osfans.trime.ime.core.TrimeInputMethodService.onCreate(TrimeInputMethodService.kt:235)
> 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4485)
> 	at android.app.ActivityThread.-$$Nest$mhandleCreateService(Unknown Source:0)
> 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2160)
> 	at android.os.Handler.dispatchMessage(Handler.java:106)
> 	at android.os.Looper.loopOnce(Looper.java:201)
> 	at android.os.Looper.loop(Looper.java:288)
> ```

这个应该不属于本 PR 带来的问题。你把部署的主题删掉了，在特定条件那当然是会炸的，因为找不到参数来初始化键盘。虽然这里有异常处理和一些边缘情况处理不完善，但是本质是因为核心文件缺失导致的。

goofyz commented:
而且的 `develop` branch 沒有此錯誤的啊

WhiredPlanck commented:
> 而且的 `develop` branch 沒有此錯誤的啊

在本 PR 以前，rime 后端应该是跑在主线程上，而且生命周期长短是不稳定的（有人抱怨时不时就能看到弹起键盘显示正在部署中）。删掉 build 文件时很有可能后端已经停机了，再次弹起键盘时又会重新开始部署建立会话。这次 PR 的带来的主要特性就是能让后端更稳定得保持运行。你看日志里有好几处 whenReady 字样，说明是在后端在线的时候出错的。在不主动杀掉应用的前提下删掉 build 就好像正在工作的电脑突然直接拔了内存，自然会崩溃的。

goofyz commented:
不，我是在主动杀掉 trime 後才刪除 `build` 資料夾的，再啟動它在死掉了 (並且不能復活)。

WhiredPlanck commented:
> 不，我是在主动杀掉 trime 後才刪除 `build` 資料夾的，再啟動它在死掉了 (並且不能復活)。

有点奇怪，我晚点排查一下

WhiredPlanck committed:
fix: forgot to require full check when manually deploy or sync user data

WhiredPlanck commented:
@goofyz 现在应该好一点了，在 app 首页手动部署一下应该能恢复

goofyz commented:
還是會出這個，一開程序就會 crash，想再部署也沒機會

```
kotlin.UninitializedPropertyAccessException: lateinit property currentSchema has not been initialized
		at com.osfans.trime.data.schema.SchemaManager.getActiveSchema(SchemaManager.kt:26)
		at com.osfans.trime.ime.symbol.TabManager.addTabHasKeys(TabManager.kt:111)
		at com.osfans.trime.ime.symbol.TabManager.refresh(TabManager.kt:52)
		at com.osfans.trime.data.theme.ThemeManager.setNormalTheme(ThemeManager.kt:70)
		at com.osfans.trime.data.theme.ThemeManager.init(ThemeManager.kt:58)
		at com.osfans.trime.data.theme.ColorManager.init(ColorManager.kt:98)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invokeSuspend(TrimeInputMethodService.kt:238)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:8)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:4)
		at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invokeSuspend(RimeDaemon.kt:59)
		at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:8)
		at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:4)
		at com.osfans.trime.core.RimeLifecycleKt.whenAtState(RimeLifecycle.kt:82)
		at com.osfans.trime.daemon.RimeDaemon$establish$1.runOnReady(RimeDaemon.kt:135)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1.invokeSuspend(TrimeInputMethodService.k
```

WhiredPlanck commented:
@goofyz 了解，我再看一看

WhiredPlanck commented:
@goofyz 请再看看有无改善

goofyz commented:
幾個問題：

1. 切換其他輸入法，再切換回來時，trime 的啟動很慢，像是要重新部署一次似的，鍵盤要很久才會出現 (5 秒以上)。此 PR 前的 trime 啟動很快，鍵盤秒彈出來。
2. 按主介面的「部署」按鈕去部署時，有時會發現"正在部署" 的畫面要等一會才顯示。追查後發現，「部署」時會執行 `DataManager.sync()`，而此 function 時快時慢，有時需要十多秒才能完成。但此 function 執行時卻不會顯示 "正在部署" 的畫面。能否將此 function 納入「正在部署」中呢？
3. 現在的 dialog style 不能轉換方案。「啟用方案」、「設定」和「其他輸入法」按鈕都不是用來切換的，點擊方案後就只是 highlight 了該方案，沒有進行切換。
![image](https://github.com/osfans/trime/assets/2428334/48706620-e3f0-48d8-8fbf-e88933bb354d)

WhiredPlanck commented:
> 1. 切換其他輸入法，再切換回來時，trime 的啟動很慢，像是要重新部署一次似的，鍵盤要很久才會出現 (5 秒以上)。此 PR 前的 trime 啟動很快，鍵盤秒彈出來。

此问题我无法复现，我怎么切速度都和这个 PR 之前差不多，需要日志进一步判断。

> 2\. 按主介面的「部署」按鈕去部署時，有時會發現"正在部署" 的畫面要等一會才顯示。追查後發現，「部署」時會執行 `DataManager.sync()`，而此 function 時快時慢，有時需要十多秒才能完成。但此 function 執行時卻不會顯示 "正在部署" 的畫面。能否將此 function 納入「正在部署」中呢？

已经纳入了，`DataManager.sync` 是写在整个启动函数之中的，而 `stateFlow` 会在启动前和后各发送正在启动和准备好的通知。而“正在部署”则分别会在正在启动和准备好时显示和消失。预计是主线程阻塞（后端的起和停是在主线程的，毕竟线程不能自己启动和关闭自己）。

> 3\. 現在的 dialog style 不能轉換方案。「啟用方案」、「設定」和「其他輸入法」按鈕都不是用來切換的，點擊方案後就只是 highlight 了該方案，沒有進行切換。

重写的对话框中执行选择方案的 CoroutineScope 没有实际执行，这个困扰我一阵子了。但我会修好。

WhiredPlanck committed:
fix: actions in dialog didn't actually launch sometimes

Requires LifecycleCoroutineScope to launch actions in dialog

WhiredPlanck commented:
@goofyz 针对上述又做了点调整。辛苦再看看有没有改善。

goofyz commented:
這個始終不定時出現，可能是 sleep mode 沒理它，一 unlock 去便出現，中間沒做任何動作。
有時死在 `onCreate()` 有時是 `recreateInputView()`。

```
TrimeInputMethodService.kt:246] kotlin.UninitializedPropertyAccessException: lateinit property activeTheme has not been initialized
		at com.osfans.trime.data.theme.ThemeManager.getActiveTheme(ThemeManager.kt:53)
		at com.osfans.trime.ime.text.TextInputManager.onCreate(TextInputManager.kt:104)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invokeSuspend(TrimeInputMethodService.kt:243)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:8)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1$1.invoke(Unknown Source:4)
		at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invokeSuspend(RimeDaemon.kt:59)
		at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:8)
		at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:4)
		at com.osfans.trime.core.RimeLifecycleKt.whenAtState(RimeLifecycle.kt:82)
		at com.osfans.trime.daemon.RimeDaemon$establish$1.runOnReady(RimeDaemon.kt:135)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onCreate$1.invokeSuspend(TrimeInputMethodService.kt:236)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
		at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
		at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
		at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
		at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
		at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
		at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
		at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
		at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
		at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
		at com.osfans.trime.ime.core.TrimeInputMethodService.onCreate(TrimeInputMethodService.kt:235)
		at android.app.ActivityThread.handleCreateService(ActivityThread.java:4485)
		at android.app.ActivityThread.-$$Nest$mhandleCreateService(Unknown Source:0)
```

```
kotlin.UninitializedPropertyAccessException: lateinit property activeTheme has not been initialized
	at com.osfans.trime.data.theme.ThemeManager.getActiveTheme(ThemeManager.kt:53)
	at com.osfans.trime.ime.core.InputView.getTheme(InputView.kt:61)
	at com.osfans.trime.ime.core.InputView.<init>(InputView.kt:89)
	at com.osfans.trime.ime.core.TrimeInputMethodService.recreateInputView(TrimeInputMethodService.kt:311)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreateInputView$1$1.invokeSuspend(TrimeInputMethodService.kt:436)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreateInputView$1$1.invoke(Unknown Source:8)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreateInputView$1$1.invoke(Unknown Source:4)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invokeSuspend(RimeDaemon.kt:59)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:8)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runOnReady$2$1.invoke(Unknown Source:4)
	at com.osfans.trime.core.RimeLifecycleKt.whenAtState(RimeLifecycle.kt:82)
	at com.osfans.trime.daemon.RimeDaemon$establish$1.runOnReady(RimeDaemon.kt:135)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onCreateInputView$1.invokeSuspend(TrimeInputMethodService.kt:435)
```

WhiredPlanck committed:
refactor: enhance handling of exception when built files are removed by user manually

WhiredPlanck commented:
@goofyz 复现了，可能还是不能方案和主题一起部署，不知道为何这样会导致 `activeTheme` 未被初始化。现在 revert 回去了。还麻烦再试试。

goofyz commented:
問題1:

在鍵盤上啟動的的 `themePicker` not work.

```
java.lang.ClassCastException: android.view.ContextThemeWrapper cannot be cast to androidx.lifecycle.LifecycleOwner
		at com.osfans.trime.ui.components.CoroutineChoiceDialog.<init>(CoroutineChoiceDialog.kt:17)
		at com.osfans.trime.ui.components.CoroutineChoiceDialog.<init>(CoroutineChoiceDialog.kt:30)
		at com.osfans.trime.ui.main.PickersKt.themePicker(Pickers.kt:15)
		at com.osfans.trime.ime.text.TextInputManager$onEvent$1.invokeSuspend(TextInputManager.kt:347)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
		at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
		at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
		at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
		at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
		at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
		at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
		at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
		at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
		at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
		at com.osfans.trime.ime.text.TextInputManager.onEvent(TextInputManager.kt:345)
		at com.osfans.trime.ime.keyboard.KeyboardView.onLongPress(KeyboardView.kt:1202)
		at com.osfans.trime.ime.keyboard.KeyboardView.openPopupIfRequired(KeyboardView.kt:1179)
		at com.osfans.trime.ime.keyboard.KeyboardView.access$openPopupIfRequired(KeyboardView.kt:65)
		at com.osfans.trime.ime.keyboard.KeyboardView$MyHandler.handleMessage(KeyboardView.kt:382)
		at android.os.Handler.dispatchMessage(Handler.java:106)
		at android.os.Looper.loopOnce(Looper.java:201)
		at android.os.Looper.loop(Looper.java:288)
		at android.app.ActivityThread.main(ActivityThread.java:7918)
		at java.lang.reflect.Method.invoke(Native Method)
		at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
		at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```


問題2:

在主畫面按「部署」的話，它會立刻顯示「完成」訊息，但實際上它還是在背景繼續 deploy 中。在部署完成前，點擊「方案」會沒有反應，部署完成後，選擇方案對話框才會跳出來。

問題3:

(不知如何復現) 在主畫面轉換主題，然後去輸入文字時，有機會出現：
```
android.view.ViewRootImpl$CalledFromWrongThreadException: Only the original thread that created a view hierarchy can touch its views. Expected: main Calling: DefaultDispatcher-worker-2
		at android.view.ViewRootImpl.checkThread(ViewRootImpl.java:9972)
		at android.view.ViewRootImpl.requestLayout(ViewRootImpl.java:2002)
		at android.view.View.requestLayout(View.java:26417)
		at android.view.View.requestLayout(View.java:26417)
		at android.view.View.requestLayout(View.java:26417)
		at android.view.View.requestLayout(View.java:26417)
		at android.view.View.requestLayout(View.java:26417)
		at android.view.View.requestLayout(View.java:26417)
		at androidx.constraintlayout.widget.ConstraintLayout.requestLayout(ConstraintLayout.java:3605)
		at android.view.View.requestLayout(View.java:26417)
		at androidx.constraintlayout.widget.ConstraintLayout.requestLayout(ConstraintLayout.java:3605)
		at android.view.View.requestLayout(View.java:26417)
		at android.view.View.requestLayout(View.java:26417)
		at android.widget.HorizontalScrollView.requestLayout(HorizontalScrollView.java:1628)
		at android.view.View.requestLayout(View.java:26417)
		at android.view.View.requestLayout(View.java:26417)
		at com.osfans.trime.ime.text.Candidate.removeHighlight(Candidate.kt:171)
		at com.osfans.trime.ime.text.Candidate.setText(Candidate.kt:135)
		at com.osfans.trime.ime.core.TrimeInputMethodService.updateComposing(TrimeInputMethodService.kt:1108)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onWindowShown$1.invokeSuspend(TrimeInputMethodService.kt:147)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onWindowShown$1.invoke(Unknown Source:8)
		at com.osfans.trime.ime.core.TrimeInputMethodService$onWindowShown$1.invoke(Unknown Source:4)
		at com.osfans.trime.daemon.RimeDaemon$establish$1$runIfReady$1$1.invokeSuspend(RimeDaemon.kt:66)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
		at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
		at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
		Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@5110d2e, Dispatchers.Default]

```
或
```

Process: com.osfans.trime.debug, PID: 23518
java.lang.NullPointerException
	at com.osfans.trime.ime.keyboard.KeyboardView.onBufferDraw(KeyboardView.kt:890)
	at com.osfans.trime.ime.keyboard.KeyboardView.invalidateKeys(KeyboardView.kt:1158)
	at com.osfans.trime.ime.keyboard.KeyboardView.invalidateComposingKeys(KeyboardView.kt:1165)
	at com.osfans.trime.ime.core.TrimeInputMethodService.updateComposing(TrimeInputMethodService.kt:1116)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onWindowShown$1.invokeSuspend(TrimeInputMethodService.kt:147)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onWindowShown$1.invoke(Unknown Source:8)
	at com.osfans.trime.ime.core.TrimeInputMethodService$onWindowShown$1.invoke(Unknown Source:4)
	at com.osfans.trime.daemon.RimeDaemon$establish$1$runIfReady$1$1.invokeSuspend(RimeDaemon.kt:66)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
```

問題 4:

感覺主畫面的 UI 反應變慢了。按「鍵盤設定」、「配置管理」之類的，它的標題會一秒變成英文的 "Keyboard Fragment", "Profile Fragment"，然後才切換去相關頁面，標題同時變回中文。

問題 5:

在某對話框輸入字碼 (e.g. "abcd") 但不按 space，去另一對話框時，該字碼會在 composition 中，有機會崩潰。

goofyz commented:
這個似乎還要試一段時間，我先 merge 我之前的 PR #1330 ，可以嗎？

WhiredPlanck commented:
> 這個似乎還要試一段時間，我先 merge 我之前的 PR #1330 ，可以嗎？

可以咧～

WhiredPlanck commented:
> 問題1:
> 
> 在鍵盤上啟動的的 `themePicker` not work.

收到，将很快修复。


> 問題2:
> 
> 在主畫面按「部署」的話，它會立刻顯示「完成」訊息，但實際上它還是在背景繼續 deploy 中。在部署完成前，點擊「方案」會沒有反應，部署完成後，選擇方案對話框才會跳出來。

这个和 glog v0.7.0 的 bug 有关，自从 librime 将 bundled 的 glog 升级到 v0.7.0 开始，其在 Android 上就怎么也不输出日志了（PC 上正常）。由于抓不到日志，原来用于捕获日志的方式误以为过程中没有问题，就直接显示完成了。这个目前临时解决办法好像只能降级。


> 問題 4:
> 
> 感覺主畫面的 UI 反應變慢了。按「鍵盤設定」、「配置管理」之類的，它的標題會一秒變成英文的 "Keyboard Fragment", "Profile Fragment"，然後才切換去相關頁面，標題同時變回中文。

这个一直都有，你可能只是最近才注意到，这应该是 Android 界面库本身的性能问题了。隔壁小企鹅也有轻微但肉眼可察觉的类似现象。

问题 3 和问题 5 有待进一步调查。

WhiredPlanck committed:
feat: make dialog use device default settings theme outside the app

WhiredPlanck committed:
fix: crash on typing after switching theme in the app settings

WhiredPlanck commented:
@goofyz 问题 1 和问题 3 已尝试修复。

Bambooin commented:
No deploying notification show up.

1. Delete build folder
2. Click the `Depoly` button
3. The main UI is blcoking
4. No `Deploying notification show up`
5. No build .bin sometime(Deploy failed?)
6. **The count of files in `build` folder is less than previous**

WhiredPlanck commented:
> 6\. The count of files in `build` folder is less than previous

I can't reproduce it. Could you show a compare?

Bambooin commented:
> 6. The count of files in build folder is less than previous

Seem the recent version has issue too, this patch is not the root cause.

goofyz commented:
切換到其他輸入法再切回來後，鍵盤佈局會變成 `default` ，而不是用輸入法設定的那個

WhiredPlanck commented:
> 切換到其他輸入法再切回來後，鍵盤佈局會變成 `default` ，而不是用輸入法設定的那個

请帮忙检查 develop 分支有没有这个问题，感觉不应该是这个 PR 造成的。

goofyz commented:
> > 切換到其他輸入法再切回來後，鍵盤佈局會變成 `default` ，而不是用輸入法設定的那個
> 
> 请帮忙检查 develop 分支有没有这个问题，感觉不应该是这个 PR 造成的。

nighty 版是正常的。
另外，nighty 版在主畫面部署時有 loading dialog 顯示的。

WhiredPlanck committed:
refactor: use native notification builder api to notify restarting

APIs from AndroidUtilCode have some issues.

WhiredPlanck commented:
> No deploying notification show up.
> 
>     1. Delete build folder
> 
>     2. Click the `Depoly` button
> 
>     3. The main UI is blcoking
> 
>     4. No `Deploying notification show up`
> 
>     5. No build .bin sometime(Deploy failed?)
> 
>     6. **The count of files in `build` folder is less than previous**

You are right. If we press the deploy once and wait for ready, no any notification nor dialog will show. But if we press the button again when it is still deploying, the notification will show up (however the dialog still won't show), it's really confusing.

WhiredPlanck committed:
fix,refactor: emit ready state in instead of after `dispatcher.start()`

`dispatcher.start()` will return immediately, making ui update too fast to be visible.

WhiredPlanck committed:
fix: librime backend didn't log to logcat

glog has an incorrect android detection since 0.7.0, so I add a definition manually to fix it.

WhiredPlanck commented:
@goofyz @Bambooin The dialog and notification showing issues should be fixed now.

Bambooin reviewed (approved):
No issue found with mannually testing so far.

WhiredPlanck commented:
@goofyz @Bambooin  If there is no other significant issues, I will merge this PR tonight since it takes a not short time to review.
